dbos/system_database.go

```diff
@@ -1772,6 +1778,20 @@ func (s *sysDB) recv(ctx context.Context, input recvInput) (any, error) {
 			s.logger.Debug("Received notification on condition variable", "payload", payload)
 		case <-time.After(timeout):
 			s.logger.Warn("Recv() timeout reached", "payload", payload, "timeout", input.Timeout)
+		case <-ctx.Done():
+			s.logger.Warn("Recv() context cancelled", "payload", payload, "cause", context.Cause(ctx))
+			return nil, ctx.Err()
+		}
+	} else {
+		// Do a "zero" sleep to record the sleep step in the workflow history
+		// This is to handle cases were consecutive Recv calls for the same key are made
+		// The key would exist and without this, the step IDs would be: 0 recv, 1 sleep, 2 recv, 4 recv, etc
+		_, err := s.sleep(ctx, sleepInput{
+			duration: 0,
+			stepID:   &sleepStepID,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to sleep before recv timeout: %w", err)
```

Comment:
I am moderately happy about this, but its the simpler and clearer: just always add a sleep step to receive calls, even if the key exists.

Otherwise we can have this kind of state:

```sql
function_id | function_name
1 | DBOS.sleep 
0 | DBOS.recv
2 | DBOS.recv
4 | DBOS.recv 
```

For a program that does consecutive receives on the same key.

Other solutions to not always do the "zero sleep" requires detecting this pattern or that we are recovering